src/Application/Actions/Donations/Confirm.php

```diff
@@ -0,0 +1,106 @@
+        }
+        \assert(is_array($requestBody));
+
+        $pamentMethodId = $requestBody['stripePaymentMethodId'];
+        \assert((is_string($pamentMethodId)));
+
+        $this->entityManager->beginTransaction();
+
+        $donation = $this->donationRepository->findAndLockOneBy(['uuid' => $args['donationId']]);
+        if (! $donation) {
+            throw new NotFoundException();
+        }
+
+        $paymentIntentId = $donation->getTransactionId();
+        $paymentMethod = $this->stripeClient->paymentMethods->retrieve($pamentMethodId);
+
+        if ($paymentMethod->type !== 'card') {
+            throw new HttpBadRequestException($request, 'Confirm endpoint only supports card payments for now');
+        }
+
+        // documented at https://stripe.com/docs/api/payment_methods/object?lang=php
+        // Contrary to what Stripes docblock says, in my testing 'brand' is strings like 'visa' or 'amex'. Not 'Visa' or
+        // 'American Express'
+        $cardBrand = $paymentMethod->card->brand;
+        \assert(is_string($cardBrand));
+
+        // two letter upper string, e.g. 'GB', 'US'.
+        $cardCountry = $paymentMethod->card->country;
+        \assert(is_string($cardCountry));
+
+        // at present if the following line was left out we would charge a wrong fee in some cases. I'm not happy with
+        // that, would like to find a way to make it so if its left out we get an error instead - either by having
+        // derive fees return a value, or making functions like Donation::getCharityFeeGross throw if called before it.
+        $this->donationRepository->deriveFees($donation, $cardBrand, $cardCountry);
+
+        $this->stripeClient->paymentIntents->update($paymentIntentId, [
+            // only setting things that may need to be updated at this point.
+            'metadata' => [
+                'stripeFeeRechargeGross' => $donation->getCharityFeeGross(),
+                'stripeFeeRechargeNet' => $donation->getCharityFee(),
+                'stripeFeeRechargeVat' => $donation->getCharityFeeVat(),
+            ],
+            // See https://stripe.com/docs/connect/destination-charges#application-fee
+            // Update the fee amount in case the final charge was from
+            // e.g. a Non EU / Amex card where fees are varied.
+            'application_fee_amount' => $donation->getAmountToDeductFractional(),
+            // Note that `on_behalf_of` is set up on create and is *not allowed* on update.
+        ]);
+
+        $this->stripeClient->paymentIntents->confirm($paymentIntentId, [
+            'payment_method' => $pamentMethodId,
+        ]);
+
+        $updatedIntent = $this->stripeClient->paymentIntents->retrieve($paymentIntentId);
+
+        $this->entityManager->flush();
+
+        return new JsonResponse([
+                'paymentIntent' => [
+                    'status' => $updatedIntent->status,
+                    'next_action' => $updatedIntent->next_action
```

Comment:
I think based on the docs that we need to include the secret here, and probably don't strictly need the `next_action` name.

```suggestion
                    'client_secret' => $updatedIntent->client_secret,
```

Reply:
Right, I thought I remembered putting next_action in because the frontend used that, but looking again it seems like it only uses status: https://github.com/thebiggive/donate-frontend/blob/55f830d9018f82a583ff56947e06e8ef93094b13/src/app/donation-start/donation-start-form/donation-start-form.component.ts#L810

What were you thinking we'd need to do with the secret on the frontend?

Of course we could just return `['paymentIntent' => $updatedIntent->toArray()]` but I feel like we should probably be intentional about which fields we send on the the client, and have them listed in the source so we can find them instead of having to look up stripe's docs.

Reply:
OK looks like on the client we basically need to copy stripe's code: 

```js
  } else if (response.status === "requires_action") {
    // Use Stripe.js to handle the required next action
    const {
      error,
      paymentIntent
    } = await stripe.handleNextAction({
      clientSecret: response.clientSecret
    });
```

Done in 06e87ca

Reply:
Yup, thanks!